mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1013,6 +1013,68 @@ struct TosaFoldConstantGreater : public TosaFoldConstantBinary<TosaFoldConstantG
   }
 };
 
+struct TosaFoldConstantBitwiseNot
+    : public TosaFoldConstantUnaryElementwise<TosaFoldConstantBitwiseNot,
+                                              BitwiseNotOp> {
+  using TosaFoldConstantUnaryElementwise<
+      TosaFoldConstantBitwiseNot,
+      BitwiseNotOp>::TosaFoldConstantUnaryElementwise;
+
+  DenseElementsAttr computeInteger(DenseElementsAttr values,
+                                   PatternRewriter &rewriter, TosaOp op) const {
+    return applyElementWise<APInt, APInt, IntegerType>(
+        values,
+        [](const APInt &val, IntegerType) {
+          return APInt(val.getBitWidth(), ~val.getZExtValue());
+        },
+        cast<IntegerType>(values.getElementType()));
+  }
+};
+
+struct TosaFoldConstantCeil
+    : public TosaFoldConstantUnaryElementwise<TosaFoldConstantCeil, CeilOp> {
+  using TosaFoldConstantUnaryElementwise<
+      TosaFoldConstantCeil, CeilOp>::TosaFoldConstantUnaryElementwise;
+
+  DenseElementsAttr computeFloat(DenseElementsAttr values,
+                                 PatternRewriter &rewriter, TosaOp op) const {
+    return applyElementWise<APFloat, APFloat, FloatType>(
+        values,
+        [](const APFloat &val, FloatType) {
+          // Compute ceil (APFloat unfortunately does not provide this function,
+          // such that we need to unpack here)
+          auto res = APFloat(std::ceil(val.convertToFloat()));
+          bool lostPrecision;
+          res.convert(val.getSemantics(), APFloat::rmNearestTiesToEven,
+                      &lostPrecision);
+          return res;
+        },
+        cast<FloatType>(values.getElementType()));
+  }
+};
+
+struct TosaFoldConstantErf
+    : public TosaFoldConstantUnaryElementwise<TosaFoldConstantErf, ErfOp> {
+  using TosaFoldConstantUnaryElementwise<
+      TosaFoldConstantErf, ErfOp>::TosaFoldConstantUnaryElementwise;
+
+  DenseElementsAttr computeFloat(DenseElementsAttr values,
+                                 PatternRewriter &rewriter, TosaOp op) const {
+    return applyElementWise<APFloat, APFloat, FloatType>(
+        values,
+        [](const APFloat &val, FloatType) {
+          // Compute ceil (APFloat unfortunately does not provide this function,
+          // such that we need to unpack here)
+          auto res = APFloat(std::erf(val.convertToFloat()));
```

Comment:
Please use `convertToDouble`  and the double overload of std::erf when val has double format,
`convertToFloat` and the float overload of std::erf when val has float or bfloat format.

For any other floating point format, it would be nice if we could do nothing and keep the constant unfolded.

I think for that you could add a customization point into `TosaFoldConstantUnaryElementwise`, like
``` bool isSupportedElementType() { return true;}``` where in `TosaFoldConstantUnaryElementwise::matchAndRewrite` you call that on the BaseClass (CRTP-style).

Then you can overwrite it here and only support BFloat, Float and Double.

Reply:
Right. So, `isSupportedElementType` is created. I also added it in `rsqrt` and `pow` folders, that were using `convertToFloat` function to get the FP value.